p4-16/psa/PSA.mdk

```diff
@@ -447,20 +469,31 @@ The list above is not intended to be exhaustive.
 
 All of the comments above apply to all types where numerical
 translation occurs between the controller and the data plane.
-Below is a complete list of such types, and an example of the
-annotation to use in your P4 program.
+Below is a complete list of such types planned for numerical
+translation by default with PSA, and have an annotation
+`p4runtime_translation` in the `psa.p4` include file:
 
-+ `PortId_t` or `PortIdInHeader_t` - annotation `@p4runtime_translation("port")`
-+ `ClassOfService_t` or `ClassOfServiceInHeader_t` - annotation `@p4runtime_translation("cos")`
++ `PortId_t` or `PortIdInHeader_t`
++ `ClassOfService_t` or `ClassOfServiceInHeader_t`
 
-For the types listed below, no numerical translation occurs, and the
+For the types listed below, no numerical translation occurs by
+default[^P4CNumericTranslationOptions].  A PSA
 data plane must support all numerical values from 0 up to the maximum
 value that it supports.  Except for `Timestamp_t` values, the number
 of values supported by the data plane need not be a power of 2.
 Controllers must have a way to discover each PSA device's maximum
 supported value for each of these types.
 
-+ `MulticastGroup_t`
+[^P4CNumericTranslationOptions]: The open source `p4c` P4 compiler is
+    planned to support an option to enable numerical translation for
+    additional types, without modifying your P4 program, nor the
+    `psa.p4` include file.  These additional types would be specified
+    by their name.
+
++ `MulticastGroup_t` - 0 is a special value that indicates no
```

Comment:
do you know why there seems to be an asymmetry between multicast and cloning? IIRC there is a boolean flag to indicate that cloning is required.

Reply:
Antonin, why do you ask such darned reasonable questions? :-)

The multicast group specialness of 0 is kind of inherited from P4_14 and v1model.  There didn't seem to be much reason to change it.

Cloning being different, via using a boolean flag to indicate that cloning should be performed for a packet, is likely my doing if I recall correctly.  It was not a conscious decision to make them different, but obviously neither was there a conscious decision to make them similar.  I can't think of any strong reason to choose one approach over the other for cloning, although perhaps consistency would have been nice here (e.g. making clone session id 0 mean no cloning, and != 0 meaning do cloning, like the multicast group id).